src/utils.ts

```diff
@@ -111,7 +156,321 @@ export const getPerpsContracts = async (
+        callData: contract.interface.encodeFunctionData(rpcFunctionName, [address]),
+      };
+    });
+    const result = await multicall.callStatic.aggregate(positionsCalls, { blockTag: block.number });
+    logger.info('Fetching positions by chunk', {
+      args: { n: batch.length, marketKeys: Array.from(marketKeysInBatch).join(',') },
+    });
+
+    for (const [i, data] of result.returnData.entries()) {
+      const { marketKey, address } = batch[i];
+      const { state } = markets[marketKey];
+
+      const { id, margin, lastPrice, size } = state.interface.decodeFunctionResult(
+        state.interface.getFunction(rpcFunctionName),
+        data
+      )[0];
+
+      // Due to how positions are updated on-chain, this may exist but size=0 (id=0) and hence not open.
+      if (id.eq(0) && size.eq(0)) {
+        continue;
+      }
+
+      const position: Position = {
+        account: address,
+        id: id.toString(),
+        size: wei(size)
+          .div(UNIT)
+          .toNumber(),
+        leverage: wei(size)
+          .abs()
+          .mul(lastPrice)
+          .div(margin)
+          .div(UNIT)
+          .toNumber(),
+        liqPrice: -1, // will be updated by keeper routine
+        liqPriceUpdatedTimestamp: block.timestamp,
+      };
+
+      ordersByMarket[marketKey] = (ordersByMarket[marketKey] ?? []).concat([position]);
+    }
+  }
+
+  return ordersByMarket;
+};
+
+// Fetch all open positions across supported markets, pinned at `block`.
+export const getOpenPositions = async (
+  markets: KeeperContracts['markets'],
+  multicall: Contract,
+  block: Block,
+  provider: providers.JsonRpcProvider
+): Promise<Record<string, Position[]>> => {
+  logger.info('Fetching on-chain positions pinned at block', {
+    args: { blockNumber: block.number },
+  });
+
+  const lengthByMarket = await getAddressLengthByMarket(
+    markets,
+    multicall,
+    block,
+    PaginationEntityType.POSITION
+  );
+
+  if (isEmpty(lengthByMarket)) {
+    return {};
+  }
+
+  // Unfortunately we cannot perform a multicall on position pagination due to `.call` and not `.delegate` on
+  // the multicall contract (addressesPage has a `onlyAssociatedContracts`). Instead, we'll simply iterate over
+  // possible values by lengths defined above.
+  //
+  // The expected output will match `getAddressesByLengths` (marketId -> address[]).
+  const addressesByMarket: Record<string, string[]> = {};
+  for (const marketKey in lengthByMarket) {
+    const { state } = markets[marketKey];
+    const associatedContractAddress = (await state.associatedContracts())[0];
+    const total = lengthByMarket[marketKey].toNumber();
+
+    const pages = getPaginatedFromAndTo(0, total, MAX_ADDRESS_PAGE_SIZE);
+    logger.info('Fetching addresses', { args: { marketKey, total } });
+
+    const getPositionAddressesCalls = pages.map(page => {
+      const { from, size } = page;
+      return state
+        .connect(provider.getSigner(associatedContractAddress))
```

Comment:
I first got confused here, but remember we did chat about this... 
So annoying not to be able to use multicall, but this is only on startup? How many position is there usually? I'm a little worried we might be rate limited...

If we run into problems it might be worth checking out: https://docs.ethers.org/v5/api/providers/other/#JsonRpcBatchProvider

Reply:
Discussed on Discord but adding it here for prosperity. 

Only on startup (but with the new pm2 config, restarts occur once every day).

As of testing this yesterday:

```
2023-06-17T04:26:21.163Z info Utils Fetched all (assumed) on-chain positions n=13933
2023-06-17T04:26:50.904Z info Utils Fetched all (open) on-chain positions n=1422
```

13.9k addresses found, only 1.4k open positions.

No issues with rate limiting when testing against Alchemy (which from exp was more aggressive than Infura).